File: storage/src/androidTest/java/pl/expensive/storage/CurrencyStorageTest.java
package pl.expensive.storage;

import android.database.sqlite.SQLiteDatabase;
import android.support.test.runner.AndroidJUnit4;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static pl.expensive.storage.DatabaseSchemaTestHelper.assertCurrencyCodeStored;

@RunWith(AndroidJUnit4.class)
public class CurrencyStorageTest {
    private CurrencyStorage storage;
    private SQLiteDatabase db;

    @Before
    public void setUp() {
        Database inMemDatabase = Injector.provideDatabase();
        db = inMemDatabase.getWritableDatabase();
        storage = new SQLiteBasedCurrencyStorage(inMemDatabase);
    }

    @Test
    public void storeSingleCurrency() {
        Currency currency = Currency.create("FAKE", "$#.##");

        storage.insert(currency);

        assertCurrencyCodeStored(db, "FAKE");
    }

    @Test(expected = IllegalStateException.class)
    public void storeDuplicates() {
        Currency currency = Currency.create("FAKE", "$#.##");
        
        storage.insert(currency);
        storage.insert(currency);
    }
}

File: storage/src/androidTest/java/pl/expensive/storage/CurrencyStorageTest.kt
package pl.expensive.storage

import android.content.SharedPreferences
import android.support.test.runner.AndroidJUnit4
import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import pl.expensive.storage.DatabaseSchemaTestHelper.assertCurrencyCodeStored

@RunWith(AndroidJUnit4::class)
class CurrencyStorageTest {

    @Test
    fun fallbackCurrencyWhenAccessingNotStoredDefaultCurrency() {
        val prefs = Injector.provideKeyValueStorage()
        clearPrefs(prefs)

        val default = getDefaultCurrency(prefs, _Seeds.CZK)

        assertThat(default).isEqualTo(_Seeds.CZK)
    }

    @Test
    fun storedDefaultCurrencyInsteadOfFallback() {
        val prefs = Injector.provideKeyValueStorage()
        clearPrefs(prefs)

        val currency = Currency("FAKE", "$#.##")
        setDefaultCurrency(currency, prefs)

        val default = getDefaultCurrency(prefs, _Seeds.CZK)

        assertThat(default).isEqualTo(currency)
    }

    private fun clearPrefs(prefs: SharedPreferences) {
        prefs.edit().clear().apply()
    }

    @Test
    fun storeSingleCurrency() {
        val currency = Currency("FAKE", "$#.##")

        val database = Injector.provideDatabase()
        insertCurrency(currency, database)

        assertCurrencyCodeStored(database.readableDatabase, "FAKE")
    }

    @Test(expected = IllegalStateException::class)
    fun storeDuplicates() {
        val currency = Currency("FAKE", "$#.##")

        val database = Injector.provideDatabase()
        insertCurrency(currency, database)
        insertCurrency(currency, database)
    }
}
